### PR TITLE
Breaking change for key-value pair serialization

### LINKED ISSUE #20898: Breaking change: PropertyNamingPolicy, PropertyNameCaseInsensitive, & Encoder options are honored when (de)serializing KeyValuePair instances with JsonSerializer
## PropertyNamingPolicy, PropertyNameCaseInsensitive, & Encoder options are honored when (de)serializing KeyValuePair instances with JsonSerializer

`JsonSerializer` will potentially serialize the "Key" and "Value" property names differently than in .NET Core 3.x/System.Text.Json 4.7.x (as influenced by the naming policy & encoder). When deserializing, the `PropertyNamingPolicy` and `PropertyNameCaseInsensitive` are honored. However, roundtripping is not broken as the literals "Key" and "Value" are special-cased to match when deserializing, to accommodate payloads that were serialized with previous versions.

### Version introduced

.NET 5.0

### Old behavior

On serialization:

The properties of `KeyValuePair` instances are always serialized as the literal "Key" and "Value", regardless of aforementioned options.

```cs
var options = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
KeyValuePair<int, int> kvp = KeyValuePair.Create(1, 1);
Console.WriteLine(JsonSerializer.Serialize(kvp, options));
// Expected: {"key":1,"value":1}
// Actual: {"Key":1,"Value":1}
```

On deserialization:

`JsonException` is thrown when the JSON property names are not precisely the "Key" and "Value" literals.

```cs
var options = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
string json = @"{""key"":1,""value"":1}";
JsonSerializer.Deserialize<KeyValuePair<int, int>>(json, options); // JsonException thrown
```

### New behavior

The aforementioned options are honored when serializing `KeyValuePair` instances.

```cs
var options = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
KeyValuePair<int, int> kvp = KeyValuePair.Create(1, 1);
Console.WriteLine(JsonSerializer.Serialize(kvp, options));
// {"key":1,"value":1}
```

On deserialization:

`PropertyNamingPolicy` and `PropertyNameCaseInsensitive` options are honored:

```cs
var options = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
string json = @"{""key"":1,""value"":1}";

KeyValuePair<int, int> kvp = JsonSerializer.Deserialize<KeyValuePair<int, int>>(json);
Console.WriteLine(kvp.Key); // 1
Console.WriteLine(kvp.Value); // 1
```

Note that the literals "Key" and "Value" are special-cased to match when deserializing, to accommodate payloads that were serialized with previous versions:

```cs
var options = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
string json = @"{""Key"":1,""Value"":1}";

KeyValuePair<int, int> kvp = JsonSerializer.Deserialize<KeyValuePair<int, int>>(json);
Console.WriteLine(kvp.Key); // 1
Console.WriteLine(kvp.Value); // 1
```

### Reason for change

Substantial customer feedback indicated that the `PropertyNamingPolicy` should be honored. For completeness, the `PropertyNameCaseInsensitive` & `Encoder` options are also honored so that `KeyValuePair<,>` instances are treated the same as any other POCO.

### Recommended action

For folks who find this change disruptive, the recommendation is to use a [custom converter](https://docs.microsoft.com/dotnet/standard/serialization/system-text-json-converters-how-to) that implements the desired semantics.

### Category

- [ ] ASP.NET Core
- [ ] C#
- [ ] Code analysis
- [ ] Core .NET libraries
- [ ] Cryptography
- [ ] Data
- [ ] Debugger
- [ ] Deployment
- [ ] Globalization
- [ ] Interop
- [ ] JIT
- [ ] LINQ
- [ ] Managed Extensibility Framework (MEF)
- [ ] MSBuild
- [ ] Networking
- [ ] Printing
- [ ] Security
- [ ] Serialization
- [ ] Visual Basic
- [ ] Windows Forms
- [ ] Windows Presentation Foundation (WPF)
- [ ] XML, XSLT

### Affected APIs

---
#### Issue metadata

* Issue type: breaking-change

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #20898.

[Preview link](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/serialization?branch=pr-en-us-21050#propertynamingpolicy-is-honored-when-serializing-and-deserializing-key-value-pairs).

### CODE REVIEW

File: includes/core-changes/serialization/5.0/options-honored-when-serializing-key-value-pairs.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+If this change is disruptive to you, you can use a [custom converter](../../../../docs/standard/serialization/system-text-json-converters-how-to.md) that implements the desired semantics.
+
+#### Category
+
+Serialization
+
+#### Affected APIs
+
+- <xref:System.Text.Json.JsonSerializer.Serialize(System.Object,System.Type,System.Text.Json.JsonSerializerOptions)?displayProperty=fullName>
```
Comment (gewarren) [inline includes/core-changes/serialization/5.0/options-honored-when-serializing-key-value-pairs.md:79]: Should I list other overloads of serialize and deserialize? Should I list the options as affected APIs?

Comment (layomia) [reply includes/core-changes/serialization/5.0/options-honored-when-serializing-key-value-pairs.md:79]: The change affects all the `Serialize`, `SerializeToUtf8Bytes`, `SerializeAsync`, `Deserialize`, and `DeserializeAsync` overloads on `JsonSerializer`. I'd expect something similar to what we had for the [deserialization constructors change](https://github.com/dotnet/docs/pull/21025/files#diff-c1fa0213bab7466b76f3fa016b9ccd1d8b7cdca9bd14075355c208852638d493R37-R38) here.

---

This change honors those options, but doesn't change their behavior. We shouldn't list them here.

**Review by layomia — APPROVED** [review]
LGTM

File: includes/core-changes/serialization/5.0/options-honored-when-serializing-key-value-pairs.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+#### Change description
+
+##### Serialization
+
+In .NET Core 3.x versions and in the 4.6.0-4.7.2 versions versions of the [System.Text.Json NuGet package](https://www.nuget.org/packages/System.Text.Json), the properties of <xref:System.Collections.Generic.KeyValuePair%602> instances are always serialized as "Key" and "Value" exactly, regardless of any <xref:System.Text.Json.JsonSerializerOptions.PropertyNamingPolicy?displayProperty=nameWithType> and <xref:System.Text.Json.JsonSerializerOptions.Encoder?displayProperty=nameWithType> options. The following code example shows how the <xref:System.Collections.Generic.KeyValuePair%602.Key> and <xref:System.Collections.Generic.KeyValuePair%602.Value> properties are *not* camel-cased after serialization, even though the specified property-naming policy dictates so.
```
Comment (gewarren) [inline includes/core-changes/serialization/5.0/options-honored-when-serializing-key-value-pairs.md:9]:
```suggestion
In .NET Core 3.x versions and in the 4.6.0-4.7.2 versions of the [System.Text.Json NuGet package](https://www.nuget.org/packages/System.Text.Json), the properties of <xref:System.Collections.Generic.KeyValuePair%602> instances are always serialized as "Key" and "Value" exactly, regardless of any <xref:System.Text.Json.JsonSerializerOptions.PropertyNamingPolicy?displayProperty=nameWithType> and <xref:System.Text.Json.JsonSerializerOptions.Encoder?displayProperty=nameWithType> options. The following code example shows how the <xref:System.Collections.Generic.KeyValuePair%602.Key> and <xref:System.Collections.Generic.KeyValuePair%602.Value> properties are *not* camel-cased after serialization, even though the specified property-naming policy dictates so.
```

**Pull request merged**